---
title: "Bulk Data Analysis - RNA Transcriptomics"
author: "Badran Elshenawy"
output: 
  html_document:
    toc: true
    toc_depth: 2
---

# Overview

-   In this practical session, we are going to analyse the data generated by a bulk RNA-Seq experiment. The aim of this session is go through the fundamental steps of a standardized DESeq2 pipeline, which are the following:

    -   **QC:** This is the first step in every bulk RNA-Seq pipeline, and it is meant to ensure the quality of the data to prevent outliers from dominating the results of subsequent analysis.

    -   **Normalisation:** This is the next step in the pipeline, which is vital to account for differences in total library size between the different samples.

    -   **Differential Expression Analysis & Visualisation:** This is the final part of the standard analysis pipeline whereby differential expression features for our condition of interest are obtained and visualized using heatmaps and volcano plots

    -   **Pathway Enrichment Analysis:** Understanding the underlying biology of our differential expression markers can only be achieved by elucidating the biological pathways for which these markers are enriched, and this is the focus of pathway enrichment analysis.

-   We further recommend the following online resources for bulk RNA-seq analysis & Pathway Enrichment Analysis:

    -   **DESEq2 Vignette:** [Analyzing RNA-seq data with DESeq2 (bioconductor.org)](http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)

    -   **ClusterProfiler Vignette:** [bioconductor.org/packages/release/bioc/html/clusterProfiler.html](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html)

# R Markdown setup

```{r setup, include=FALSE}
root.directory <- dirname(rstudioapi::getActiveDocumentContext()$path)
knitr::opts_chunk$set(echo = TRUE, warning = FALSE)
knitr::opts_knit$set(root.dir = root.directory)

dir.create(path = "Write/Figures/", recursive = T)
dir.create(path = "Write/Tables/", recursive = T)

options(stringsAsFactors = FALSE)
```

# Packages

-   In this tutorial, we are going to use two primary packages:

    -   **`Tidyverse`:** This is a very popular system of packages in R that facilitates common data science operations involving data wrangling and visualisations. If you would like to learn more about the tidyverse in general, please check [**here**](https://www.r-bloggers.com/2020/06/a-very-short-introduction-to-tidyverse/). For more comprehensive coverage of the system, please check [**here**](https://r4ds.had.co.nz/).

    -   **`DESeq2`:** This is the state--of-art package for analysis of bulk cell RNA-Seq data, and this is what we will be using here for the processing and and manipulation of bulk data. For more information about additional functionality that the package offers, please check the documentation (easily accessible through R) as well as the [**official vignette**](http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html).

    -   `EnhancedVolcano:` This is a very popular package for the visualisation of bulk RNA-Seq results as Volcano plots. For furhter information about the package, please check [**here**](https://bioconductor.org/packages/devel/bioc/vignettes/EnhancedVolcano/inst/doc/EnhancedVolcano.html).

    -   **`clusterProfiler:`** This is a very popular package for pathway enrichment analysis. clusterProfiler can use different pathway enrichment algorithms (over-representation analysis and gene set enrichment analysis) and queries gene ontology terms as well as KEGG functional pathways. For more information about ClusterProfiler, check the package documentation (easily accessible through R) as well as the [**official vignette**](http://yulab-smu.top/clusterProfiler-book/).

-   The final command below sets the directory this file is in to be the active directory for this session. This is crucial for the ***read*** and ***write*** functions below to work effectively since they use relative rather than absolute filepaths (if you are not sure what the difference between relative and absolute filepaths is, please check [**here**](https://www.codingrooms.com/blog/file-paths#:~:text=A%20relative%20path%20describes%20the,regularly%20use%20relative%20file%20paths.)).

```{r Packages, warning=FALSE}
# Packages
suppressMessages(library(tidyverse))
suppressMessages(library(DESeq2))
suppressMessages(library(clusterProfiler))
suppressMessages(library(EnhancedVolcano))
suppressMessages(library(pheatmap))
suppressMessages(library(org.Mm.eg.db))


#Setting Working Directory
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
```

# 1. Data Import

## 1.1. Expression Matrix Import

-   The first step is data import. Data is stored in tsv format and will be imported using the `read_tsv()` function.

```{r Data-Import}
expression_matrix <- read_tsv(file = "../../course_data/Bulk RNA-Seq Module/all_gene_counts_before_filter(Manual_Relabel).tsv") %>%
  column_to_rownames("ensembl_gene_id") %>% 
  dplyr::select("Ctrl1", "Ctrl2", "Ctrl3", "Scr1","Scr2", "Scr3")
```

## 1.2. Metadata Creation

-   Here we create the metadata dataframe for our expression matrix; this is crucial for subsequent instantiation of the DESeqDataSet object later in the DESeq2 pipeline.

-   Sanity checks below are meant to ensure that the rownames of the metadata are identical to the colnames of the expression matrix in terms of spelling and order. This is important for DESeq2 to work properly.

```{r Metadata}
# Metadata
experiment_metadata <- data.frame(row.names = colnames(expression_matrix), 
                                  Condition = c(rep("Ctrl",3), rep("Scr",3)),
                                  Replicate = c(1,2,3,1,2,3))

experiment_metadata <- experiment_metadata %>% 
  dplyr::mutate(Condition = factor(Condition, levels = c("Ctrl", "Scr")),
                Replicate = factor(Replicate, levels = c(1,2,3)))

# Sanity Checks
all(rownames(experiment_metadata) %in% colnames(expression_matrix))
all(rownames(experiment_metadata) == colnames(expression_matrix))
```

# 2. DESeq2 Object Instantiation

-   Here we instantiate the DESeqDataSet object, which will store the expression matrix as well as the metadata.

```{r DESeq2-Object-Instantiation}
# DESeqDataSet Instantiation
dds <- DESeqDataSetFromMatrix(countData = expression_matrix,
                       colData = experiment_metadata,
                       design = ~Replicate + Condition)

# Exploration
head(colData(dds)) # sample metadata
head(rowData(dds)) # feature metadata
head(counts(dds)) # count matrix assay
```

# 3. QC

## 3.1. Genes QC

-   This filtration step will remove all features that do not have non-zero accounts in at least 10 cells.

-   This filtration step is essential since lowly-expressed genes can come up as variable genes in subsequent steps and will skew the results of the analysis if not removed early.

```{r Gene-QC}
dim(dds) # 20689 features
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
dim(dds) # 14977 features
```

-   **Note:** It's important to filter first the cells and then the genes.

## 3.2. Sample QC

### 3.2.1. PCA Plots

-   The `rlog()` transformation below is needed for PCA since PCA requires log-transformed data as input.

```{r PCA, fig.height=6, fig.width=6}
# rlog=======
dds_rlog <- rlog(object = dds, blind = T)

# PCA=======
plotPCA(dds_rlog, intgroup = "Condition")
ggsave(filename = "Write/Figures/Condition PCA.tiff", 
       height = 10, 
       width = 10)

plotPCA(dds_rlog, intgroup = "Replicate")
ggsave(filename = "Write/Figures/Replicate PCA.tiff", 
       height = 10, 
       width = 10)
```

### 3.2.2. Correlation Heatmaps

```{r Correlation-Heatmaps, fig.height=6, fig.width=8}
#Heatmaps=======
#Sample Correlation Heatmap
dds_rlog_cor <- cor(assay(dds_rlog))
tiff("Write/Figures/Correlation Heatmap.tiff", res = 200, units = "in", height = 10, width = 10)
pheatmap(dds_rlog_cor, annotation = dplyr::select(data.frame(dds_rlog@colData), Condition, Replicate))
dev.off()

# Sample Distance Heatmap
sampleDists <- dist(t(assay(dds_rlog)))
sampleDistsMatrix <- as.matrix(sampleDists)

tiff("Write/Figures/Sample Distances Heatmap.tiff", res = 200, units = "in", height = 10, width = 10)
pheatmap(sampleDistsMatrix, annotation = dplyr::select(data.frame(dds_rlog@colData), Condition))
dev.off()
```

# 4. Differential Expression Analysis (DEA)

```{r DESeq2-Results}
# DESeq Run
dds <- DESeq(dds)

# Results Table
results_table <- results(
  # dds object
  dds, 
  # Experimental Contrast
  contrast = c("Condition", "Scr", "Ctrl"), 
  # Significance Threshold
  alpha = 0.05)

# Useful Summary
DESeq2::summary(results_table)

# Dataframe Creation
results_table <- data.frame(results_table) %>% rownames_to_column("ensgenes")

# Annotation (Gene Symbols)
results_table$gene_symbol <- AnnotationDbi::mapIds(x = org.Mm.eg.db,
                                    keys = results_table$ensgenes,
                                    keytype = "ENSEMBL", 
                                    column = "SYMBOL",
                                    multiVals = "first")
results_table$entrezid <- AnnotationDbi::mapIds(x = org.Mm.eg.db,
                                    keys = results_table$ensgenes,
                                    keytype = "ENSEMBL", 
                                    column = "ENTREZID",
                                    multiVals = "first")

# Removing NAs
results_table <- results_table %>% dplyr::filter(!is.na(entrezid))

# Saving Results Table
write_csv(x = results_table, 
          file = "Write/Tables/DESeq2 Results.csv")
```
  
```{r Volcano-Plot, fig.height=10, fig.width=15}  
# Volcano Plot
EnhancedVolcano(toptable = results_table, 
                lab = results_table$gene_symbol, 
                x = "log2FoldChange", 
                y = "padj")
ggsave(filename = "Write/Figures/EnhancedVolcano.tiff",
       height = 10, 
       width = 15)
```

# 5. Pathway Enrichment Analysis (PEA)

## 5.1. Gene Set Enrichment Analysis (GO Terms)

```{r GSEA-GO}
# Wrangling
results_table <- results_table %>% arrange(desc(log2FoldChange))
gene_list <- structure(results_table$log2FoldChange, names = results_table$entrezid)

# GSEA GO Terms (Biological Pathways)
gse_go_res <- gseGO(geneList = gene_list,
              OrgDb = org.Mm.eg.db,
              ont = "BP",
              keyType = "ENTREZID", 
              eps = 0) %>% setReadable(OrgDb = org.Mm.eg.db,
                                                    keyType = "ENTREZID")

# Saving
write_csv(gse_go_res@result, file = "Write/Tables/GSEA GO Results.csv")
# Exploration of Results
view(gse_go_res@result)
```

## 5.2. Visualisation

```{r PEA-Viz, fig.height=6, fig.width=8}
# dotplot
enrichplot::dotplot(object = gse_go_res)
ggsave(filename = "Write/Figures/GSEA GO DotPlot.tiff",
       height = 10,
       width = 10)
```
